src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+    }
+
+    #endregion
+
+    #region Constructor Overload Resolution Tests
+
+    [Fact]
+    public void WithElement_OverloadResolution_ExactMatch()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public string ConstructorUsed { get; }
+                
+                public MyList(int capacity) : base(capacity)
+                {
+                    ConstructorUsed = "int";
+                }
+                
+                public MyList(long capacity) : base((int)capacity)
+                {
+                    ConstructorUsed = "long";
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list1 = [with(10), 1];
+                    MyList<int> list2 = [with(10L), 2];
+                    Console.WriteLine($"{list1.ConstructorUsed},{list2.ConstructorUsed}");
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("int,long"));
+    }
+
+    [Fact]
+    public void WithElement_OverloadResolution_Ambiguous()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public MyList(int value) : base() { }
+                public MyList(long value) : base() { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    short s = 10;
+                    MyList<int> list = [with(s)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void WithElement_OverloadResolution_BestMatch()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public string ConstructorUsed { get; }
+                
+                public MyList(object value) : base()
+                {
+                    ConstructorUsed = "object";
+                }
+                
+                public MyList(int value) : base()
+                {
+                    ConstructorUsed = "int";
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(42), 1];
+                    Console.WriteLine(list.ConstructorUsed);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("int"));
+    }
+
+    [Fact]
+    public void WithElement_NoMatchingConstructor()
```

Comment:
Consider adding tests around user-defined conversions as well, both singularly and in comparison with built-in conversions. For example, an `int` value with both `long` and `CanBeConvertedFromInt` types available.